README.md
# Ruby Paint [<img src="https://badge.fury.io/rb/paint.svg" />](https://badge.fury.io/rb/paint) [<img src="https://github.com/janlelis/paint/workflows/Test/badge.svg" />](https://github.com/janlelis/paint/actions?query=workflow%3ATest)

Paint creates terminal colors and effects for you. It combines the strengths of **term-ansicolor**, **rainbow**, and similar projects into a simple to use, however still flexible terminal colors gem with no core extensions by default.

Supported Rubies: **3.1**, **3.0**, **2.7**

Unsupported, but might still work: **2.6**, **2.5**, **2.4**, **2.3**, **2.2**, **2.1**, **2.0**, **1.9**

## Features

* No string extensions (suitable for library development)
* Simple API
* Faster than other terminal color gems ([as of July 2022](https://gist.github.com/janlelis/91413b9295c81ee873dc))
* Supports *true color* or 256 colors (for capable terminals)
* Allows you to set any terminal effects
* `Paint.mode`: Fall-back modes for terminals with less colors, supported modes:
  * 0xFFFFFF (= 16777215) colors (*true color*)
  * 256 colors (palette)
  * 16 colors (only ANSI colors, combined with bright effect)
  * 8 colors (only ANSI colors)
  * 0 colors (no colors / deactivate)

## Paint 2.0 | True Color Support

Starting with **Paint 2.0**, *true color* mode is the new default mode, since most major terminals now support 24bit colors. If it happens to not work in your setup:

- Manually set `Paint.mode = 256` at the beginning of your code
- Please [open a new issue](https://github.com/janlelis/paint/issues/new) so we can figure out how to blacklist the terminal used

## Setup

Add to `Gemfile`:

```ruby
gem 'paint'
```

and run `bundle install`.

In Ruby do:

```ruby
require 'paint'
```

## Usage

The only method you need is: `Paint.[]`

The first argument given to `Paint.[]` is the string to colorize (if the object is not a string, `to_s` will be called on it). The other arguments describe how to modify/colorize the string. Let's learn by example:

```ruby
Paint['Ruby', :red]           # Sets ANSI color red
Paint['Ruby', :red, :bright]  # Also applies bright/bold effect
Paint['Ruby', :bright, :red]  # Does the same as above
Paint['Ruby', :red, :bright, :underline] # Effects can often be combined
Paint['Ruby', :red, :blue]    # The second color you define is for background
Paint['Ruby', nil, :blue]     # Pass a nil before a color to ignore foreground and only set background color
Paint['Ruby', [100, 255, 5]]  # You can define RGB colors. Depending on your terminal, this will create
                              # a "true color" or map to 256/16/8 colors.
Paint['Ruby', "gold", "snow"] # Paint supports rgb.txt color names, note that the arguments are strings
                              # (:yellow != "yellow")!
Paint['Ruby', "#123456"]      # HTML like definitions are possible
Paint['Ruby', "fff"]          # Another HTML hex definition
Paint['Ruby', :inverse]       # Swaps fore- and background
Paint['Ruby', :italic, :encircle, :rapid_blink, :overline] # Probably not supported effects
Paint['Ruby']                 # Don't pass any argument and the string will not be changed
```

When you pass multiple colors, the first one is taken as foreground color and the second one defines the background color, every following color will be ignored. To only change the background color, you have to pass a `nil` first. Effects can be passed in any order.

[You can find more examples in the specs.](https://github.com/janlelis/paint/blob/master/spec/paint_spec.rb)

[List of rgb.txt colors.](https://en.wikipedia.org/wiki/X11_color_names#Color_name_chart)

## Windows Support

For ANSI support in Windows OS, you can use [ansicon](https://github.com/adoxa/ansicon) or [ConEmu](https://conemu.github.io/) or [WSL](https://docs.microsoft.com/en-us/windows/wsl/install-win10).

## `Paint.mode`

You can choose between five ways to use `Paint.[]` by setting `Paint.mode` to one of the following:

* **0xFFFFFF**: Use 16777215 *true colors*
* **256**:      Use the 256 colors palette
* **16**:       Use the eight ANSI colors (combined with bright effect)
* **8**:        Use the eight ANSI colors
* **0**:        Don't colorize at all

Paint tries to automatically detect the proper value your terminal is capable of, please [open an issue](https://github.com/janlelis/paint/issues/new) if `Paint.detect_mode` yields a wrong value for you.

`Paint.detect_mode` will return 0 if the [NO_COLOR environment variable is set](https://github.com/jcs/no_color/).

## More Details About Terminal Colors and Effects

<img src="https://pbs.twimg.com/media/ENyLvgVXUAgeDTn.jpg" />

Terminal colors/effects get created by [ANSI escape sequences](https://en.wikipedia.org/wiki/ANSI_escape_code). These are strings that look like this: `\e[X;X;X;X;X]m` where X are integers with some meaning. For example, `0` means *reset*, `31` means *red foreground* and `41` stands for red background. When you tell **Paint** to use one of the eight ANSI base colors as foreground color, it just inserts a number between `30` and `37` into the sequence. The following colors are available:

* `:black`
* `:red`
* `:green`
* `:yellow`
* `:blue`
* `:magenta`
* `:cyan`
* `:white`, `:gray`
* (`:default`)

When combined with the `:bright` (= `:bold`) effect, the color in the terminal emulator often differs a little bit, thus it is possible to represent 16 colors.

Through special sequences it's also possible to set 256-colors, or even 16777215 colors, instead of only the 8 ANSI ones. However, this is not supported by all terminals. Paint automatically translates given RGB colors to a suitable color of the supported color spectrum.

When using the `Paint.[]` method, Paint wraps the given string between the calculated escape sequence and an reset sequence (`"\e[0m"`). You can get the raw escape sequence by using the `Paint.color` method.

### Effects

See [en.wikipedia.org/wiki/ANSI_escape_code](https://en.wikipedia.org/wiki/ANSI_escape_code) for a more detailed discussion:

#### Often supported

    0) :reset, :nothing
    1) :bright, :bold
    4) :underline
    7) :inverse, :negative
    8) :conceal, :hide
    22) :clean
    24) :underline_off
    26) :inverse_off, :positive
    27) :conceal_off, :show, :reveal

#### Not widely supported

    2) :faint
    3) :italic
    5) :blink, :slow_blink
    6) :rapid_blink
    9) :crossed, :crossed_out
    10) :default_font, :font0
    11-19) :font1, :font2, :font3, :font4, :font5, :font6, :font7, :font8, :font9
    20) :fraktur
    21) :bright_off, :bold_off, :double_underline
    23) :italic_off, :fraktur_off
    25) :blink_off
    29) :crossed_off, :crossed_out_off
    51) :frame
    52) :encircle
    53) :overline
    54) :frame_off, :encircle_off
    55) :overline_off

## Substitution & Nesting

From time to time, you might find yourself in a situation where you want to colorize a substring differently from the rest of the string. Paint supports this via a simple templating approach using the `%` method with an array argument. Use the `%{var}` notation within a string, and pass the template variables as a hash:

```ruby
Paint%['Yellow string with a %{blue_text} in it', :yellow,
  blue_text: ["blue text", :blue]
]
# => "\e[33mYellow string with a \e[34mblue text\e[0m\e[33m in it\e[0m"
```

Please note that the resulting ASCII escape sequence can be quite verbose since it restores the parent context after the substitution.

## Utilities

The `Paint.random` method generates a random ANSI color you can pass into `Paint.[]`:

```ruby
Paint['Ruby', Paint.random]        # Get one of eight random ANSI foreground colors
Paint['Ruby', Paint.random(true)]  # Get one of eight random ANSI background colors
```

Another helper method is `Paint.unpaint`, which removes any ANSI colors:

```ruby
Paint.unpaint( Paint['Ruby', :red, :bright] ).should == 'Ruby'
```

You can get a `p` like alternative for calling `puts Paint.[]`:

```ruby
require 'paint/pa'
pa "Ruby", :red, :underline  # same as puts Paint["Ruby", :red, :underline]
```

## Advanced Usage: Shortcuts

There is an extension gem available which allows you to define custom color definitions, which you can reuse later. See [SHORTCUTS.md](https://github.com/janlelis/paint/blob/main/SHORTCUTS.md) for documentation. This is completely optional.

## J-_-L

Copyright (c) 2011-2021 Jan Lelis <https://janlelis.com>, released under the
MIT license.

Thank you to [rainbow](https://github.com/sickill/rainbow) and [term-ansicolor](https://github.com/flori/term-ansicolor) for ideas and inspiration. Also, a lot of thanks to all the [contributors](https://github.com/janlelis/paint/contributors)!


CHANGELOG.md
# CHANGELOG

### 2.3.0

#### Paint.%() now returns different escape sequences!

* Fix #30: Color/effect nesting restores the parent level correctly. More
           verbose ASCII escape sequences are generated to ensure this.

#### Other Fixes

* Fix #31: Make paint work with frozen strings enabled
    * Also make paint library itself frozen
* Fix #29: Update dev dependencies

### 2.2.1

* Explicitly set mac's Terminal.app to 256 colors only, fixes #28

### 2.2.0

* Support NO_COLOR environment variable, implements #26 (see no-color.org)

### 2.1.1

* Blacklist True Color support for urxvt, fixes #25

### 2.1.0

* Set True Color as default mode on more terminals, patch by @smoochbot

### 2.0.3

* Add `gray` alias for `white` color, patch by @AlexWayfer

### 2.0.2

* Remove `gunzip` deprecation warning

### 2.0.1

*   Fix nested substitutions, patch by @mildmojo

### 2.0.0
#### Major Changes

*   New default color mode `0xFFFFFF`: 24bit - true color. If this breaks your code, add `Paint.mode = 256` to the beginning of your code
*   New `Paint%[]` API: Substitution mechanism for nested color strings

#### Minor Changes

*   Smaller gem size (compress RGB color name data)
*   Remove `Paint.update_rgb_colors` and `Paint.rainbow`
*   Internal method `.hex` renamed to `.rgb_hex` and does not take "#" prefixed strings anymore
*   Minor refactorings and documentation updates

### 1.0.1

*   Fix case of string arguments getting mutated (see gh#14)


### 1.0.0

*   Improved performance
*   Option for :random colors removed (see readme)
*   Separate Paint::SHORTCUTS into extra gem
*   Drop support for Ruby 1 (inoffically still support 1.9.3)


### 0.9.0

*   Don't colorize strings via shortcuts when Paint.mode == 0
*   Freeze bundled ascii color data


### 0.8.7

*   Fix caching bug for random ansi color


### 0.8.6

*   Add missing require 'rbconfig' and travis test everything


### 0.8.5

*   Support 256 color on windows' ConEmu


### 0.8.4

*   Fix post-install message unicode


### 0.8.3

*   Paint.[] also accepts uppercased hex strings (gh#2)
*   Performance tweaks (thanks to murphy) (gh#4, #5)
    *   API change: deactivate colorizing with Paint.mode = 0


### 0.8.2

*   Paint.[] with only a single string argument does not colorize the string
    anymore, but returns the plain string
*   New pseudo color :random - returns a random ansi color


### 0.8.1

*   Improve rgb function with better gray scale values
*   Add Paint.mode:
    *   Set to 0 to deactivate colorizing
    *   Set to 16 or 8 and all color generation methods will generate simple
        ansi colors
    *   Set to 256 for 256 color support
    *   Tries to automatically detect your terminal's features
*   Minor changes


### 0.8.0

*   Initial release



lib/paint/constants.rb
# frozen_string_literal: true

module Paint
  # Clears all colors
  NOTHING = "\033[0m"

  # Number of possible colors in TRUE COLOR mode
  TRUE_COLOR = 0xFFFFFF

  # Basic colors (often, the color differs when using the bright effect)
  # Final color will be 30 + value for foreground and 40 + value for background
  ANSI_COLORS = {
    :black   => 0,
    :red     => 1,
    :green   => 2,
    :yellow  => 3,
    :blue    => 4,
    :magenta => 5,
    :cyan    => 6,
    :white   => 7, :gray => 7,
    :default => 9,
  }.freeze

  ANSI_COLORS_FOREGROUND = {
    :black   => 30,
    :red     => 31,
    :green   => 32,
    :yellow  => 33,
    :blue    => 34,
    :magenta => 35,
    :cyan    => 36,
    :white   => 37, :gray => 37,
    :default => 39,
  }.freeze

  ANSI_COLORS_BACKGROUND = {
    :black   => 40,
    :red     => 41,
    :green   => 42,
    :yellow  => 43,
    :blue    => 44,
    :magenta => 45,
    :cyan    => 46,
    :white   => 47, :gray => 47,
    :default => 49,
  }.freeze

  # Terminal effects - most of them are not supported ;)
  # See https://en.wikipedia.org/wiki/ANSI_escape_code
  ANSI_EFFECTS = {
    :reset         => 0,  :nothing         => 0,  # usually supported
    :bright        => 1,  :bold            => 1,  # usually supported
    :faint         => 2,
    :italic        => 3,
    :underline     => 4,                          # usually supported
    :blink         => 5,  :slow_blink      => 5,
    :rapid_blink   => 6,
    :inverse       => 7,  :swap            => 7,  # usually supported
    :conceal       => 8,  :hide            => 9,
    :default_font  => 10,
    :font0 => 10, :font1 => 11, :font2 => 12, :font3 => 13, :font4 => 14,
    :font5 => 15, :font6 => 16, :font7 => 17, :font8 => 18, :font9 => 19,
    :fraktur       => 20,
    :bright_off    => 21, :bold_off        => 21, :double_underline => 21,
    :clean         => 22,
    :italic_off    => 23, :fraktur_off     => 23,
    :underline_off => 24,
    :blink_off     => 25,
    :inverse_off   => 26, :positive        => 26,
    :conceal_off   => 27, :show            => 27, :reveal           => 27,
    :crossed_off   => 29, :crossed_out_off => 29,
    :frame         => 51,
    :encircle      => 52,
    :overline      => 53,
    :frame_off     => 54, :encircle_off    => 54,
    :overline_off  => 55,
  }.freeze

  # A list of color names for standard ansi colors, needed for 16/8 color fallback mode
  # See https://en.wikipedia.org/wiki/ANSI_escape_code#Colors
  RGB_COLORS_ANSI = {
    :black   => [  0,   0,   0],
    :red     => [205,   0,   0],
    :green   => [  0, 205,   0],
    :yellow  => [205, 205,   0],
    :blue    => [  0,   0, 238],
    :magenta => [205,   0, 205],
    :cyan    => [  0, 205, 205],
    :white   => [229, 229, 229], :gray => [229, 229, 229],
  }.each { |k, v| v.freeze }.freeze

  # A list of color names for standard bright ansi colors, needed for 16 color fallback mode
  # See https://en.wikipedia.org/wiki/ANSI_escape_code#Colors
  RGB_COLORS_ANSI_BRIGHT = {
    :black   => [127, 127, 127],
    :red     => [255,   0,   0],
    :green   => [  0, 255,   0],
    :yellow  => [255, 255,   0],
    :blue    => [ 92,  92, 255],
    :magenta => [255,   0, 255],
    :cyan    => [  0, 255, 255],
    :white   => [255, 255, 255], :gray => [255, 255, 255],
  }.each { |k, v| v.freeze }.freeze
end
